File: src/lib/events.ts
// Single event channel from the engine (aos-core AppEvent, tag = "type").

import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { Classification, MonitoringState } from "./ipc";

export type AppEvent =
  | { type: "ACTIVITY_CHANGED"; app_name: string; window_title: string; classification: Classification }
  | { type: "USER_IDLE" }
  | { type: "USER_ACTIVE" }
  | { type: "FOCUS_STARTED"; commitment_id: number }
  | { type: "FOCUS_ENDED"; commitment_id: number }
  | { type: "COMMITMENT_CHANGED"; commitment_id: number | null }
  | { type: "DISTRACTION_WARNING"; distracted_secs: number }
  | { type: "DISTRACTION_DETECTED"; distracted_secs: number; app_name: string; window_title: string }
  | { type: "DISTRACTION_RESOLVED"; recovery_secs: number | null }
  | { type: "CHECKIN_DUE"; checkin_id: number }
  | { type: "CHECKIN_ANSWERED"; checkin_id: number }
  | { type: "PRIORITY_CHANGE_REQUESTED"; commitment_id: number | null }
  | { type: "BLOCKED_FLOW_REQUESTED"; commitment_id: number | null }
  | { type: "BREAK_STARTED"; ends_at: number }
  | { type: "BREAK_ENDED" }
  | { type: "TASK_COMPLETED"; task_id: number }
  | { type: "DAY_LOCKED"; plan_id: number }
  | { type: "DAY_ENDED"; plan_id: number }
  | { type: "INTERVIEW_DUE" }
  | { type: "REVIEW_DUE" }
  | { type: "MONITORING_STATUS"; state: MonitoringState }
  | { type: "SESSIONS_UPDATED" }
  | { type: "SCORES_UPDATED" };

export function onAppEvent(handler: (event: AppEvent) => void): Promise<UnlistenFn> {
  return listen<AppEvent>("app-event", (e) => handler(e.payload));
}

File: crates/aos-core/src/events.rs
//! Internal event model (spec §31). The engine emits these; the UI renders
//! them. Serialized as a tagged union on a single event channel.

use serde::{Deserialize, Serialize};

use crate::types::Classification;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum AppEvent {
    ActivityChanged {
        app_name: String,
        window_title: String,
        classification: Classification,
    },
    UserIdle,
    UserActive,
    FocusStarted {
        commitment_id: i64,
    },
    FocusEnded {
        commitment_id: i64,
    },
    CommitmentChanged {
        commitment_id: Option<i64>,
    },
    DistractionWarning {
        distracted_secs: i64,
    },
    DistractionDetected {
        distracted_secs: i64,
        app_name: String,
        window_title: String,
    },
    DistractionResolved {
        recovery_secs: Option<i64>,
    },
    CheckinDue {
        checkin_id: i64,
    },
    CheckinAnswered {
        checkin_id: i64,
    },
    PriorityChangeRequested {
        commitment_id: Option<i64>,
    },
    BlockedFlowRequested {
        commitment_id: Option<i64>,
    },
    BreakStarted {
        ends_at: i64,
    },
    BreakEnded,
    TaskCompleted {
        task_id: i64,
    },
    DayLocked {
        plan_id: i64,
    },
    DayEnded {
        plan_id: i64,
    },
    InterviewDue,
    ReviewDue,
    MonitoringStatus {
        state: MonitoringState,
    },
    /// Activity sessions changed; timeline views should refetch.
    SessionsUpdated,
    /// Scores changed; scorecards should refetch.
    ScoresUpdated,
}

/// Monitoring status shown in the UI at all times (spec §41).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MonitoringState {
    Active,
    Paused,
    PermissionRequired,
    /// Demo simulation driving the monitor instead of the OS.
    Demo,
}
